load_sample_data.py

```diff
@@ -0,0 +1,469 @@
+        EOF"""
+    output = compose_exec(UPSTREAM_DB_SERVICE_NAME, bash_input)
+    logging.info(f"Table '{name}' copied to '{target_name}'.")
+    return output
+
+
+def run_just(recipe: str, argv: list[str]) -> subprocess.CompletedProcess:
+    """
+    Run the given ``just`` recipe with the given arguments.
+
+    :param recipe: the name of the ``just`` recipe to invoke
+    :param argv: the list of arguments to pass after the ``just`` recipe
+    :return: the process obtained from the ``subprocess.run`` command
+    """
+
+    try:
+        logging.debug(f"just {recipe} {' '.join(argv)}")
+        proc = subprocess.run(
+            ["just", recipe, *argv],
+            check=True,
+            capture_output=True,
+            text=True,
+        )
+        logging.debug(f"Output: {proc.stdout}")
+        return proc
+    except subprocess.CalledProcessError as exc:
+        logging.error("Just call failed.")
+        logging.error(f"STDOUT: {exc.stdout}")
+        logging.error(f"STDERR: {exc.stderr}")
+        raise
+
+
+def get_actual_providers() -> list[str]:
+    """
+    Get the list of all distinct providers mentioned in the sample data.
+
+    :return: the list of unique providers in the sample data
+    """
+
+    providers = set()
+    for media_type in MEDIA_TYPES:
+        sample_file_path = f"./sample_data/sample_{media_type}.csv"
+        with open(sample_file_path, "r") as sample_file:
+            reader = csv.DictReader(sample_file)
+            for row in reader:
+                providers.add(row["provider"])
+    return list(providers)
+
+
+#########
+# Steps #
+#########
+
+
+def run_migrations():
+    """
+    Run all migrations for the API.
+    """
+
+    logging.info("Executing migrations...")
+    bash_input = "python manage.py migrate --noinput"
+    compose_exec(WEB_SERVICE_NAME, bash_input)
+    logging.info("Migrations executed.")
+
+
+def create_users(names: list[str]):
+    """
+    Create users with the given usernames in the API database. The password for the
+    users is always set to "deploy". Users that already exist will not be recreated.
+    """
+
+    logging.info("Creating users...")
+    bash_input = f"""{DJ_SHELL} <<EOF
+        from django.contrib.auth.models import User
+        usernames = {names}
+        for username in usernames:
+            if User.objects.filter(username=username).exists():
+                print(f'User {{username}} already exists')
+                continue
+            if username == 'deploy':
+                user = User.objects.create_superuser(
+                    username, f'{{username}}@example.com', 'deploy'
+                )
+            else:
+                user = User.objects.create_user(
+                    username, f'{{username}}@example.com', 'deploy'
+                )
+                user.save()
+        EOF"""
+    compose_exec(WEB_SERVICE_NAME, bash_input)
+    logging.info("Users created.")
+
+
+def backup_table(media_type: MediaType):
```

Comment:
This step is new to me, do we need to backup test data? 🤔

Reply:
This doesn't back up the test data, it only keeps a copy of the `image` and `audio` tables so that we can run the ingestion script again idempotently.

During ingestion, the ingestion server modifies the tables in the downstream DB, which means that if we run the `load_sample_data` script again, it will fail. This step allows us to use the initial tables (as created by Django) as the starting point every time we run this script.

Reply:
1. Django migration creates `image` in downstream.
2. We copy the `image` table to `image-template`.
3. `image-template` is used to build the `image-view` table in upstream.
4. Data is loaded into `image-view` from the CSV.
5. Data refresh overwrites downstream `image` with upstream `image-view`.

Since `image` has changed in step 5 we cannot use it to create `image-view` again. By having it backed up in step 2 we ensure that step 3-5 can be repeated again and again, allowing us to run `just init` on an existing init-ed setup.